File: src/lib/routineWeek.ts
import { format, subDays, isValid } from 'date-fns';
import type { WorkoutDay, WorkoutLog } from '../store/useAppStore';

export const WEEKDAY_KEYS = ['lun', 'mar', 'mie', 'jue', 'vie', 'sab', 'dom'] as const;

export const WEEKDAY_LABELS = [
  { key: 'lun', short: 'Lun', full: 'Lunes' },
  { key: 'mar', short: 'Mar', full: 'Martes' },
  { key: 'mie', short: 'Mie', full: 'Miercoles' },
  { key: 'jue', short: 'Jue', full: 'Jueves' },
  { key: 'vie', short: 'Vie', full: 'Viernes' },
  { key: 'sab', short: 'Sab', full: 'Sabado' },
  { key: 'dom', short: 'Dom', full: 'Domingo' },
] as const;

function normalizeText(value: string) {
  return value
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .trim();
}

function getWeekdayIndexFromName(dayName: string): number {
  const normalized = normalizeText(dayName);

  if (normalized.includes('lun')) return 0;
  if (normalized.includes('mar')) return 1;
  if (normalized.includes('mie') || normalized.includes('mié')) return 2;
  if (normalized.includes('jue')) return 3;
  if (normalized.includes('vie')) return 4;
  if (normalized.includes('sab') || normalized.includes('sáb')) return 5;
  if (normalized.includes('dom')) return 6;

  return -1;
}

export function getMondayFirstIndex(date: Date) {
  const day = date.getDay();
  return day === 0 ? 6 : day - 1;
}

/** Weeks start on Monday (ISO-8601). */
export const WEEK_STARTS_ON_MONDAY = 1 as const;

/** Maximum number of past days to look back when computing streaks. */
const MAX_STREAK_DAYS = 365;

export function mapRoutineByWeekday(routine: WorkoutDay[]) {
  const mapped: Array<WorkoutDay | null> = new Array<WorkoutDay | null>(7).fill(null);
  const unresolved: WorkoutDay[] = [];

  routine.forEach((day) => {
    const idx = getWeekdayIndexFromName(day.day);
    if (idx >= 0 && mapped[idx] === null) {
      mapped[idx] = day;
      return;
    }
    unresolved.push(day);
  });

  unresolved.forEach((day) => {
    const freeIdx = mapped.findIndex((entry) => entry === null);
    if (freeIdx >= 0) {
      mapped[freeIdx] = day;
    }
  });

  return mapped;
}

/**
 * Computes the current training streak, skipping rest days.
 * A "rest day" is any weekday that has no exercises assigned in the routine.
 * Only scheduled workout days that were missed (no logs) break the streak.
 * Today's workout is not penalised if not yet done.
 */
export function computeSmartStreak(logs: WorkoutLog[], routine: WorkoutDay[]): number {
  if (logs.length === 0) return 0;

  const routineByDay = mapRoutineByWeekday(routine);
  const workoutWeekdays = new Set<number>(
    routineByDay
      .map((entry, index) => (entry ? index : -1))
      .filter((index) => index >= 0),
  );

  const dateSet = new Set(
    logs
      .filter((log) => isValid(new Date(log.date)))
      .map((log) => format(new Date(log.date), 'yyyy-MM-dd')),
  );

  const todayStr = format(new Date(), 'yyyy-MM-dd');
  let streak = 0;
  let cursor = new Date();

  for (let i = 0; i < MAX_STREAK_DAYS; i++) {
    const dateStr = format(cursor, 'yyyy-MM-dd');
    const weekdayIndex = getMondayFirstIndex(cursor);
    // When no routine is defined every calendar day counts (original behaviour).
    const isWorkoutDay = workoutWeekdays.size === 0 || workoutWeekdays.has(weekdayIndex);

    if (isWorkoutDay) {
      if (dateSet.has(dateStr)) {
        streak += 1;
      } else if (dateStr === todayStr) {
        // Today's session not yet completed — don't penalise.
      } else {
        // Missed a past workout day — streak ends.
        break;
      }
    }
    // Rest day: skip without breaking streak.
    cursor = subDays(cursor, 1);
  }

  return streak;
}

File: voltbody-android/app/src/main/java/com/voltbody/app/domain/usecase/RoutineWeek.kt
package com.voltbody.app.domain.usecase

import com.voltbody.app.domain.model.WorkoutDay
import com.voltbody.app.domain.model.WorkoutLog
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatter

// ── Routine week helpers (ported from routineWeek.ts) ─────────────────────────

val WEEKDAY_LABELS = listOf(
    Triple("lun", "Lun", "Lunes"),
    Triple("mar", "Mar", "Martes"),
    Triple("mie", "Mie", "Miércoles"),
    Triple("jue", "Jue", "Jueves"),
    Triple("vie", "Vie", "Viernes"),
    Triple("sab", "Sab", "Sábado"),
    Triple("dom", "Dom", "Domingo")
)

fun getMondayFirstIndex(date: LocalDate): Int {
    return when (date.dayOfWeek) {
        DayOfWeek.MONDAY -> 0
        DayOfWeek.TUESDAY -> 1
        DayOfWeek.WEDNESDAY -> 2
        DayOfWeek.THURSDAY -> 3
        DayOfWeek.FRIDAY -> 4
        DayOfWeek.SATURDAY -> 5
        DayOfWeek.SUNDAY -> 6
    }
}

private fun normalizeText(value: String): String =
    value.lowercase()
        .normalize()
        .trim()

private fun String.normalize(): String {
    val normalized = java.text.Normalizer.normalize(this, java.text.Normalizer.Form.NFD)
    return normalized.replace(Regex("\\p{InCombiningDiacriticalMarks}+"), "")
}

private fun getWeekdayIndexFromName(dayName: String): Int {
    val n = normalizeText(dayName)
    return when {
        n.contains("lun") -> 0
        n.contains("mar") -> 1
        n.contains("mie") -> 2
        n.contains("jue") -> 3
        n.contains("vie") -> 4
        n.contains("sab") -> 5
        n.contains("dom") -> 6
        else -> -1
    }
}

fun mapRoutineByWeekday(routine: List<WorkoutDay>): Array<WorkoutDay?> {
    val mapped = arrayOfNulls<WorkoutDay>(7)
    val unresolved = mutableListOf<WorkoutDay>()

    routine.forEach { day ->
        val idx = getWeekdayIndexFromName(day.day)
        if (idx >= 0 && mapped[idx] == null) {
            mapped[idx] = day
        } else {
            unresolved.add(day)
        }
    }

    unresolved.forEach { day ->
        val freeIdx = mapped.indexOfFirst { it == null }
        if (freeIdx >= 0) mapped[freeIdx] = day
    }

    return mapped
}

private const val MAX_STREAK_DAYS = 365
private val DATE_FMT = DateTimeFormatter.ISO_LOCAL_DATE

fun computeSmartStreak(logs: List<WorkoutLog>, routine: List<WorkoutDay>): Int {
    if (logs.isEmpty()) return 0

    val routineByDay = mapRoutineByWeekday(routine)
    val workoutWeekdays = routineByDay
        .mapIndexed { i, d -> if (d != null) i else -1 }
        .filter { it >= 0 }
        .toSet()

    val dateSet = logs
        .mapNotNull { runCatching { LocalDate.parse(it.date.take(10), DATE_FMT) }.getOrNull() }
        .map { it.format(DATE_FMT) }
        .toSet()

    val today = LocalDate.now()
    val todayStr = today.format(DATE_FMT)
    var streak = 0
    var cursor = today

    repeat(MAX_STREAK_DAYS) {
        val dateStr = cursor.format(DATE_FMT)
        val weekdayIndex = getMondayFirstIndex(cursor)
        val isWorkoutDay = workoutWeekdays.isEmpty() || workoutWeekdays.contains(weekdayIndex)

        if (isWorkoutDay) {
            when {
                dateSet.contains(dateStr) -> streak++
                dateStr == todayStr -> { /* today not done yet — don't break */ }
                else -> return streak // missed past workout day
            }
        }
        cursor = cursor.minusDays(1)
    }

    return streak
}
